2455 add protected areas section

yaguzmang opened the pull request:
Solves #2455 by:

Refactoring:
-  Dividing the logic to handle map layers: separating the layers handlers in different files, so they are more modular and easier to maintain.
- Renaming "layers" to "forestLayers" since there will be new non-forest layers (this is also done in the redux reducers).

Adding new logic and UI:
- Added a new hook, redux state management, and UI components for the protected areas.


TODO: Find a way to render the WDPA layer. After spending quite some time trying to get the WDPA layers to work, it seems like it is not possible the way it is set up now.

As per the [documentation](https://developers.google.com/earth-engine/guides/featureview_overview): "The process for visualizing a FeatureView asset using the Google Maps API is a four part process: obtain a tiles key using ee.data.getFeatureViewTilesKey, pass it to an instance of FeatureViewTileSource, construct an ee.layer.ImageOverlay [...]" at the moment we call `ee.data.getFeatureViewTilesKey` in the backend and then we send the response to the frontend. In the mapController we then try to pass it to a new `FeatureViewTileSource`, but due to the typing that is lost when sending the object from backend to frontend, the `FeatureViewTileSource` constructor doesn't recognise the object and throws an error. Unfortunately, there is no casting available for this kind of object.  



![Large GIF (862x484)](https://user-images.githubusercontent.com/41337901/229308747-d2018b99-50e8-497e-8208-3284a7a60b53.gif)

yaguzmang committed:
Divide the logic to handle map layers

Dividing the logic to handle different kinds of layers,
since there will be more in the future. Having them in
separate files is better for readability and makes the
code more modular (e.g. you could now just handle forest
layers on their own, without having to handle all layers).

yaguzmang committed:
Rename "layers" to "forestLayers" & add protected area layers

yaguzmang committed:
Rename redux reducer & add ProtectedAreasOptions

Renaming redux reducers to be more specific, considering
that there will be more layers that are not part of the
forest layers. Also, adding the types for the protected
areas layers.

yaguzmang committed:
Add redux state management for Protected Areas

yaguzmang committed:
Add UI logic for the Protected Areas

yaguzmang committed:
Prevent custom asset id input from reseting to empty

minotogna commented on src/client/pages/Geo/GeoMap/GeoMapMenuData/MapVisualizerPanel/ProtectedAreaLayerSection/CustomAssetControl/CustomAssetControl.scss:
```diff
@@ -0,0 +1,137 @@
+      cursor: default;
+    }
+  }
+
```
Any chance to use the global color variables ?

yaguzmang replied:
Yes, fixed this and a bug in the new commit.

minotogna commented on src/client/store/ui/geo/slice.ts:
```diff
@@ -174,9 +182,57 @@ export const geoSlice = createSlice({
+    },
+    setCustomProtectedAreaAssetId: (state, { payload }: PayloadAction<string>) => {
+      state.protectedAreasOptions.customAssetId = payload
+    },
```
This file is becoming too big , that means there's either common logic to refactor or we should split this into several files ... what do you think ?

yaguzmang replied:
Yes, I agree it is getting too big. The logic for the layers is similar but not always the same, so generalizing that logic would end up in a lot of extra code handling edge cases. I think it should better be split. Is it okay to do it in the upcoming Burned Areas PR?

minotogna replied:
let's think how to refactor this later on?

yaguzmang replied:
Yes, splitting it into files is a good start, but there should be a better way to handle the layers. We can take a look into that later.

minotogna replied:
well, the best way would be to avoid duplicated code.

mergify[bot] committed:
Merge branch 'development' into 2455-add-protected-areas-section

yaguzmang committed:
Clear unused scss classes and fix ui bug in input

Clearing the unsed styling, and using global variables instead.
Also fixed a bug causing the method trim() to be called on a null
value, which is the initial value of the customAssetId.

mergify[bot] committed:
Merge branch 'development' into 2455-add-protected-areas-section

mergify[bot] committed:
Merge branch 'development' into 2455-add-protected-areas-section

mergify[bot] committed:
Merge branch 'development' into 2455-add-protected-areas-section

mergify[bot] committed:
Merge branch 'development' into 2455-add-protected-areas-section

mergify[bot] committed:
Merge branch 'development' into 2455-add-protected-areas-section

mergify[bot] committed:
Merge branch 'development' into 2455-add-protected-areas-section
